<h1>Ece Alptekin</h1>
<h1>24156</h1>

In [1]:
import pandas as pd
import numpy as np

<h2>Import data</h2>

In [2]:
df = pd.read_csv("coding_potential.csv") 
df

#ID  ORF-integrity  ORF-coverage  Instability      T2        C0  \
0      0.162071            1.0      0.114583    56.694764   576.0  0.000281   
1      0.431156            1.0      0.086155    41.790444  1287.0  0.002011   
2      0.874798            1.0      0.081715    40.573639  1083.0  0.004850   
3      0.853372            1.0      0.081498    41.800543  1164.0  0.004402   
4      0.789583            1.0      0.080668    42.827804  1137.0  0.004170   
...         ...            ...           ...          ...     ...       ...   
57518  0.000000            0.0      0.064516     0.000000     0.0  0.042553   
57519  0.200000            1.0      0.048544    50.101667    21.0  0.038462   
57520  0.461538            1.0      0.066667    55.985185    84.0  0.005525   
57521  0.253125            1.0      0.044025    98.788462    81.0  0.012539   
57522  0.382550            1.0      0.054422   112.877778    57.0  0.006757   

           PI  ORF-length        AC         T0  ...        C4        AG  \
0      0.5279    0.369828  0.700535   7.877136  ...  0.993808  0.995215   
1      0.5395    0.471515  0.717828   8.553284  ...  0.997654  0.998324   
2      0.9489    0.427648  0.752627   8.643616  ...  0.995150  0.988682   
3      0.9297    0.468819  0.765957   8.255188  ...  1.000000  0.994864   
4      0.8437    0.468381  0.762335   8.588196  ...  1.000000  0.995136   
...       ...         ...       ...        ...  ...       ...       ...   
57518  0.7840    0.585106  0.765957   0.000000  ...  0.925532  0.957447   
57519  0.9510    0.192308  0.778846   8.344666  ...  0.951923  0.971154   
57520  1.0046    0.629834  0.690608   3.925598  ...  0.994475  1.000000   
57521  0.9168    0.438871  0.699060  12.004700  ...  0.987461  0.996865   
57522  0.5492    0.472973  0.797297   8.000305  ...  0.891892  1.000000   

        Fickett        A2        T4         C         G         A         T  \
0     -0.704188  0.613566  0.152590  0.107545  0.353223  0.247959  0.244019   
1      0.368458  0.520107  0.115320  0.088166  0.232239  0.210456  0.262064   
2      0.496667  0.482619  0.100324  0.111650  0.243331  0.219887  0.247373   
3      0.393540  0.442406  0.110866  0.107930  0.222304  0.225238  0.256787   
4      0.525661  0.455177  0.109179  0.104312  0.223767  0.220987  0.261293   
...         ...       ...       ...       ...       ...       ...       ...   
57518  0.000000  0.457447  0.064516  0.086022  0.255319  0.138298  0.255319   
57519 -0.816667  0.500000  0.077670  0.126214  0.230769  0.221154  0.269231   
57520  0.048148  0.558011  0.138889  0.050000  0.292818  0.232044  0.292818   
57521 -0.438462  0.561129  0.125786  0.100629  0.231975  0.216301  0.253918   
57522 -0.138889  0.513514  0.156463  0.095238  0.270270  0.263514  0.277027   

       label  
0          1  
1          1  
2          1  
3          1  
4          1  
...      ...  
57518      0  
57519      0  
57520      0  
57521      0  
57522      0  

[57523 rows x 39 columns]

<h4>1. How many rows and columns does the coding potential.csv have?</h4>
57523 rows, 39 columns

In [3]:
X = np.array(df.iloc[:, :-1].values)  
y = np.array(df.iloc[:, -1].values) 

<h2>Split the data into train and test sets</h2>

In [4]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X, y)

In [5]:
print(np.shape(X_train))
print(np.shape(y_train))
print(np.shape(X_test))
print(np.shape(y_test))

(43142, 38)
(43142,)
(14381, 38)
(14381,)


<h4>2. When you split the data into segments, how many samples (rows) are in the training and test data?</h4>
There are 43142 rows in the training data. There are 14381 rows in the test data.

<h2>Train a random forest model</h2>

In [20]:
from sklearn.ensemble import RandomForestClassifier
clf=RandomForestClassifier()
clf.fit(X_train,y_train)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='auto',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, n_estimators=100,
                       n_jobs=None, oob_score=False, random_state=None,
                       verbose=0, warm_start=False)

<h2>Evaluate the model</h2>

<h2>Predicting train set results</h2>

In [21]:
ypred_train=clf.predict(X_train)

<h2>Confusion Matrix</h2>

In [22]:
from sklearn.metrics import confusion_matrix
print("Train Confusion Matrix:")
print(confusion_matrix(y_train, ypred_train))

Train Confusion Matrix:
[[18174     3]
 [    4 24961]]


<h2>Classification Report</h2>

In [23]:
from sklearn.metrics import classification_report
print("Train Classification Report:")
print(classification_report(y_train, ypred_train))

Train Classification Report:
              precision    recall  f1-score   support

           0       1.00      1.00      1.00     18177
           1       1.00      1.00      1.00     24965

    accuracy                           1.00     43142
   macro avg       1.00      1.00      1.00     43142
weighted avg       1.00      1.00      1.00     43142



<h4>3. Report the confusion matrices and classification reports for the training data.</h4>
In training data, we have the precision 1.00 and recall 1.00. The model predicts very well. The f1-score is 1.00, which is the best value. As we look at the confusion matrix, there are 3 False Positives and 4 False Negatives. We conclude that the accuracy is very high. 

<h2>Predicting test set results</h2>

In [24]:
ypred_test=clf.predict(X_test)

<h2>Confusion Matrix</h2>

In [25]:
from sklearn.metrics import confusion_matrix
print("Test Confusion Matrix:")
print(confusion_matrix(y_test, ypred_test))

Test Confusion Matrix:
[[5854  132]
 [ 160 8235]]


<h2>Classification Report</h2>

In [28]:
from sklearn.metrics import classification_report
print("Test Classification Report:")
print(classification_report(y_test, ypred_test))

Test Classification Report:
              precision    recall  f1-score   support

           0       0.97      0.98      0.98      5986
           1       0.98      0.98      0.98      8395

    accuracy                           0.98     14381
   macro avg       0.98      0.98      0.98     14381
weighted avg       0.98      0.98      0.98     14381



<h4>3. Report the confusion matrices and classification reports for the test data.</h4>
In test data, we have the precision 0.97 and recall 0.98. The model predicts good. The f1-score is 0.98, which is very close to 1. As we look at the confusion matrix, there are 132 False Positives and 160 False Negatives. We conclude that the model predicts good but the model performs better when predicting the trainig data, since The precision, recall and f1-score of the training data is higher than the test data's corresponding precision, recall and f1-score.

<h4>4. Explain what the confusion matrix describes.</h4>
A confusion matrix is a table, which describes the performance of a classification model on a set of training or test data for which the true values are known. Confusion matrix shows the number of True Positives, False Positives(The system predicted true but the value was false), False Negatives(The system predicted false but the value was true) and True Negatives.

<h4>5. What are precision, recall and F1-score?</h4>
Precision is the percentage of the items that the system labeled as positive and that are in fact positive. It is equal to True Positive/(True Positive+False Positive). Recall or sensitivity is the percentage of the items actually present in the input that were correctly identified by the system. It is equal to True Positive/(True Positive+False Negative). F-measure can be weighted to favor Precision or Recall. It is equal to ((1+B^2)PR)/((B^2)P+R). 𝛽 = 1 is the F1-measure.

<h4>6. Does the classifier perform better when predicting the training data or the test data targets?</h4>
The classifier performs better when predicting the training data.

<h4>7. Why is there a difference between the training and test prediction accuracies?</h4>
Because the test data is never seen before by the model until the prediction process. The model was trained by the training data, therefore the model is used to the training data and the training data's accuracy is higher than the test data's accuracy. The model models the training data too well, therefore overfitting occurs.